### PR TITLE
Add methods to select hyperparameters for LogisticRegression and RandomForestClassifier models

### PR DESCRIPTION
**Pull request by kschelonka**
Use [hyperopt](http://hyperopt.github.io/hyperopt/) to tune models. For each hyperparameter "trial", sample the hyperparameter space. Use the sampled params to train a model with k-fold cross validation. Compute the average score of the optimization metric over the folds. Return the best parameters found and retrain the model on the training set using those parameters (optionally refit to include the test set). Compute metrics on the train and test sets and log them to mlflow.

The design of the classes are easy to use for interactive, notebook-based data exploration/model prototyping. By including intelligent defaults in the model-specific tuner classes, they should be able to be effectively used in scripts as long as the search space is well-defined and number of iterations is sufficiently large (for RandomForestClassifier -- LogisticRegression doesn't really have a lot of tunable parameters).

Notable changes compared to previous version:

0. Add the tuning capabilities
1. Update arguments for entry into the module. All added arguments have defaults so are not required.
2. Compute all metrics before entering into the mlflow log block.
3. Potentially refit model on all data (train + test)
4. Update the mlflow tagging and add metrics
5. Update the way we store confusion matrix (use the labeled keys "TP", "TF", etc.)
6. Update some of the online_training tests that were failing on my machine during schema validation (from files that don't exist)

Edit: Additional Validation
I set up our cloud infrastructure and got it running. Note that we will need to update the docker image that the stack points to. We should probably automate this in circleCI on merge to master.

I ran a model with 1000 hyperopt trials using our very small feature set. 
`run_id= f887d2be2bad43b68143f8d7787178a0`
All tags and params logged in mlflow as expected.
Here's the model artifact: `s3://prod.croissant-artifacts.alleninstitute.org/f887d2be2bad43b68143f8d7787178a0/artifacts/trained_model.joblib`
Here are the results: 
```
INFO:TrainClassifier:Considered 1000 trials over 1437.4327154159546 seconds. Best Score: 0.8294377932372532
Best Params: {'bootstrap': False, 'max_depth': None, 'max_features': 'sqrt', 'min_samples_leaf': 10, 'min_samples_split': 2, 'n_estimators': 101, 'n_jobs': 1, 'oob_score': False, 'random_state': 42, 'verbose': False}
```
Screen Shot 2020-08-11 at 12 29 18 PM

### CODE REVIEW

**Comment by njmei** [comment]
@kschelonka Looks like your circleci config is missing a backslash `\`

https://github.com/AllenInstitute/croissant/blob/57f724fce2bc58ffec73b34dafae74fb35cffb48/.circleci/config.yml#L76

Which is why the `mlflow-cli` test is failing.

It's weird though since on master that backslash is not missing...

File: test/test_train.py [hunk 1]
```diff
@@ -103,82 +83,254 @@ def test_mlflow_log_classifier(tmp_path, mock_classifier,
 
     mlflow run reads the tracking URI from the env variable
     MLFLOW_TRACKING_URI
-
+        experiment_name = "myexperiment"
```
Comment (njmei) [inline test/test_train.py:86]: Are these lines supposed to be in the docstring?

File: test/test_train.py [hunk 2]
```diff
@@ -103,82 +83,254 @@ def test_mlflow_log_classifier(tmp_path, mock_classifier,
...
+    with pytest.raises(ValueError, match=r"Must have binary labels"):
+        train._binary_confusion_dict(y_true, y_pred)
+
+
+@pytest.mark.parametrize(
+    "tags,metrics,confusion_matrix",
+    [
+        (
+            {"never": "gonna", "give": "you"},
+            {"up": 0.867, "never": 0.5309},
```
Comment (njmei) [inline test/test_train.py:217]: 🤣

File: test/test_train.py (hunk 1 above)
Comment (kschelonka) [reply test/test_train.py:86]: Thanks I'll take a look, maybe was a merge issue

Comment (kschelonka) [reply test/test_train.py:86]: Aha, this was an accident when updating tests. Thanks.

File: croissant/train.py [hunk 3]
```diff
@@ -1,169 +1,673 @@
...
+        y: Iterable
+            True labels for the data to train the model.
+        classifier: Type[BaseEstimator]
+            A scikit-learn classifier class. Should be able to be
+            instantiated from params in the param space.
+        pipeline: Pipeline
+            An optional Pipeline. Will use to preprocess data, and
+            append the classifier to the end of the Pipeline.
+            Will not search for hyperparameters to tune in the Pipeline.
+        param_space: Iterable
```
Comment (njmei) [inline croissant/train.py:256]: Should the `param_space` be a `dict`?

At least that's what it looks to be:
https://github.com/hyperopt/hyperopt-sklearn/blob/b9fe2d778a09f3b59c4458a058f21c970a2cec22/hpsklearn/components.py#L752

Comment (kschelonka) [reply croissant/train.py:256]: Good catch, yes I made an error there.

File: croissant/train.py [hunk 4]
```diff
@@ -1,169 +1,673 @@
...
+        start_time = time.time()
+        estimator = classifier(**params)
+        if pipeline is None:
+            new_pipe = Pipeline(steps=[("classifier", estimator)])
+        else:
+            new_pipe = clone(pipeline)
+            new_pipe.steps.append(("classifier", estimator))
+        loss = np.mean(
+            cross_val_score(new_pipe, X, y, scoring=scorer, cv=n_splits))
+        if not low_good:
```
Comment (njmei) [inline croissant/train.py:161]: Since `scorer` is a possible param, will this do the right thing if someone sets `greater_is_better=False` or `greater_is_better=True`?

See: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html

**Comment by kschelonka** [comment]
Thanks @njmei . Looks like I broke the tests with what I just thought were typing updates. I'll fix that and see if I can ping someone to take another look if you want.

File: croissant/train.py (hunk 4 above)
Comment (kschelonka) [reply croissant/train.py:161]: I mentioned it in the docstring but I can be more clear. Scikit-learn follows a convention that all scorers follow the convention that higher values are good.

Comment (kschelonka) [reply croissant/train.py:161]: But thinking more on this, I actually don't think it's an issue. I provided options for both `fmin` and `fmax` out of convenience. If a low score is good, run `fmin` to optimize. Otherwise run `fmax`.

Comment (kschelonka) [reply croissant/train.py:161]: Sorry for the comment spam. The ultimate answer is that the user will have to use the correct functions if they are creating a custom scorer. I think it's fairly obvious what to do (I mean I wrote it), but let me know if it can be more clear.

Comment (njmei) [reply croissant/train.py:161]: Sounds reasonable to me!

File: croissant/train.py [hunk 5]
```diff
@@ -1,169 +1,697 @@
...
+def train_classifier(model: str,
+                     training_data_path: str,
+                     test_data_path: str,
+                     scorer: Union[str, Callable],
+                     max_iter: int,
+                     optimizer: str,
+                     test_metrics: Optional[List[str]] = None,
+                     n_folds: int = 5,
+                     seed: int = 42,
+                     refit: bool = True) -> Pipeline:
```
Comment (djkapner) [inline croissant/train.py:402]: return hint here and docstring don't reflect the tuple you're returning

File: croissant/train.py [hunk 6]
```diff
@@ -1,169 +1,697 @@
...
 
 
 def mlflow_log_classifier(training_data_path: str,
                           test_data_path: str,
-                          clf: GridSearchCV) -> str:
-    """Logs a classifier with mlflow
+                          clf: GridSearchCV,
+                          cv_scores: tuple,
+                          metrics: Optional[dict] = None,
+                          confusion_matrix: Optional[np.ndarray] = None,
```
Comment (djkapner) [inline croissant/train.py:507]: is this not a `dict` coming back from `_binary_confusion_dict`?

File: test/test_train.py [hunk 7]
```diff
@@ -1,15 +1,33 @@
...
+from functools import partial
+from urllib.parse import urlparse
 from unittest.mock import MagicMock
+
+import pytest
 import mlflow
-import mlflow.sklearn
-import joblib
-from functools import partial
+# import mlflow.sklearn
```
Comment (djkapner) [inline test/test_train.py:10]: delete line?

File: test/test_train.py [hunk 8]
```diff
@@ -188,11 +333,150 @@ def test_ClassifierTrainer(train_data, test_data, tmp_path, monkeypatch):
...
+@pytest.mark.parametrize(
+    "param_space, expected_params", [
+        (None, ["bootstrap", "max_depth", "max_features", "min_samples_leaf",
+                "min_samples_split", "n_estimators", "oob_score"]),
+        ({"max_depth":
+          hp.choice("max_depth", [1, 5, 10])}, ["max_depth"])
+    ]
+)
+@pytest.mark.parametrize(
+    "pipeline,n_splits,scorer,low_good",
```
Comment (djkapner) [inline test/test_train.py:427]: nit, spaces

Comment (kschelonka) [reply test/test_train.py:427]: I agree spaces look nicer, but the pytest examples on their site don't have any so I was just following that style. \*shrug\*

**Comment by kschelonka** [comment]
Good eyes from both, working on the type details. I started up a stack to show that it will work on Fargate. Nothing was spun up so I have to get everything working. I will add that validation.

File: test/test_train.py [hunk 9]
```diff
@@ -188,11 +332,150 @@ def test_ClassifierTrainer(train_data, test_data, tmp_path, monkeypatch):
...
+)
+def test_LogisticRegressionTuner_fmin_fmax(
+        X_y, param_space, pipeline, n_splits, scorer, low_good,
+        expected_params):
+    X, y = X_y
+    tuner = train.LogisticRegressionTuner(
+        X, y, pipeline, param_space=param_space,
+        n_splits=n_splits, scorer=scorer)
+    if low_good:
+        best_params = tuner.fmin(2)
```
Comment (isaak-willett) [inline test/test_train.py:474]: I'm curious as to the rational for not parameterizing the max_evals parameter for this function call.

Comment (kschelonka) [reply test/test_train.py:474]: I chose the minimum number of trials that made sense so that the tests could run quickly.
